fix: Trigger correct notifications when moving directories (#865)

stogle opened the pull request:
This PR attempts to fix missing notifications from `FileSystemWatcherMock` on Windows, when moving directories into or out of the watched path, as described in #865.

On Windows, moving a directory from a location outside the watched path to a location inside the watched path should trigger a Created event. Similarly, moving a directory from a location inside the watched path to a location outside the watched path should trigger a Deleted event. On other platforms, these moves should trigger a Rename event instead.

stogle commented on Source/Testably.Abstractions.Testing/FileSystem/FileSystemWatcherMock.cs:
```diff
@@ -353,17 +353,27 @@ private bool MatchesFilter(ChangeDescription changeDescription)
-				return changeDescription.ChangeType == WatcherChangeTypes.Renamed &&
-				       changeDescription.OldPath?.StartsWith(fullPath,
-					       _fileSystem.Execute.StringComparisonMode) == true;
+				if (!(changeDescription.ChangeType == WatcherChangeTypes.Renamed &&
+				      changeDescription.OldPath?.StartsWith(fullPath,
+					      _fileSystem.Execute.StringComparisonMode) == true))
+				{
+					return false;
+				}
```
When the `ChangeType` is `Renamed`, either `changeDescription.Path` or `changeDescription.OldPath` needs to be in the watcher's path. If neither of them are we can return false. If one of them is we still need to consider `NotifyFilters` and the `Filters` collection, so we can't return yet.

stogle commented on Source/Testably.Abstractions.Testing/FileSystem/FileSystemWatcherMock.cs:
```diff
@@ -353,17 +353,27 @@ private bool MatchesFilter(ChangeDescription changeDescription)
 			_fileSystem.Execute.StringComparisonMode))
 		{
-			return false;
+			if (!(changeDescription.ChangeType == WatcherChangeTypes.Renamed &&
+			      string.Equals(
+				      _fileSystem.Execute.Path.GetDirectoryName(changeDescription.OldPath),
+				      fullPath,
+				      _fileSystem.Execute.StringComparisonMode)))
+			{
+				return false;
+			}
```
Equivalent logic for when `IncludeSubdirectories` is false.

stogle commented on Source/Testably.Abstractions.Testing/FileSystem/FileSystemWatcherMock.cs:
```diff
@@ -538,28 +548,12 @@ private string TransformPathAndName(
 
 	private void TriggerRenameNotification(ChangeDescription item)
 	{
-		if (!item.Path.StartsWith(Path, _fileSystem.Execute.StringComparisonMode) &&
-		    item.OldPath != null)
-		{
-			Deleted?.Invoke(this, ToFileSystemEventArgs(
-				WatcherChangeTypes.Deleted, item.OldPath, item.OldName));
-		}
-		else if (_fileSystem.Execute.IsWindows)
```
Also added in #866 but again I didn't find it necessary. It seems that raising Created and Deleted notifications on renames is a Windows-only behaviour?

stogle commented on Source/Testably.Abstractions.Testing/FileSystem/FileSystemWatcherMock.cs:
```diff
@@ -568,10 +562,39 @@ private void TriggerRenameNotification(ChangeDescription item)
+				string fullPath = _fileSystem.Execute.Path.GetFullPath(Path);
+				if (IncludeSubdirectories)
+				{
+					if (item.OldPath.StartsWith(fullPath, _fileSystem.Execute.StringComparisonMode))
+					{
+						Deleted?.Invoke(this, ToFileSystemEventArgs(
+							WatcherChangeTypes.Deleted, item.OldPath, item.OldName));
+					}
+					if (item.Path.StartsWith(fullPath, _fileSystem.Execute.StringComparisonMode))
+					{
+						Created?.Invoke(this, ToFileSystemEventArgs(
+							WatcherChangeTypes.Created, item.Path, item.Name));
+					}
```
Deleted gets invoked if `OldPath` matches the watcher's path. Created gets invoked if `Path` matches the watcher's path.

stogle commented on Tests/Testably.Abstractions.Testing.Tests/FileSystem/FileSystemWatcherMockTests.cs:
```diff
@@ -299,7 +299,7 @@ public async Task RenamedEventArgs_ShouldUseDirectorySeparatorFromSimulatedFileS
 
 			using IFileSystemWatcher fileSystemWatcher =
-				fileSystem.FileSystemWatcher.New(fileSystem.Path.GetFullPath(parentDirectory));
+				fileSystem.FileSystemWatcher.New(parentDirectory);
```
This was added in #866 but I didn't find it necessary. Maybe I'm missing something?

vbreuss replied:
This was added on purpose to make the tests succeed after the following observation:
>During testing of #866, I stumbled upon the case that you initialize the file system watcher with a relative directory (e.g. "watched") instead of one with a full path and the returned paths were also relative. Currently the tests fail on the real file system:
> <img width="1478" height="287" alt="Image" src="https://github.com/user-attachments/assets/46336279-b528-463f-9ba0-9702020393b5" />

stogle commented on Tests/Testably.Abstractions.Tests/FileSystem/FileSystemWatcher/NotifyFiltersTests.cs:
```diff
@@ -613,7 +613,7 @@ public async Task NotifyFilter_MoveFile_ShouldTriggerChangedEventOnNotifyFilters
 	[Theory]
 	[InlineAutoData(NotifyFilters.DirectoryName)]
-	public async Task NotifyFilter_MoveDirectory_ShouldTriggerChangedEventOnNotifyFilters(
+	public async Task NotifyFilter_MoveDirectory_ShouldTriggerRenamedEventOnNotifyFilters(
```
Correcting the event that should be triggered.

stogle commented on Tests/Testably.Abstractions.Tests/FileSystem/FileSystemWatcher/NotifyFiltersTests.cs:
```diff
@@ -654,11 +654,13 @@ public async Task NotifyFilter_MoveDirectory_ShouldTriggerChangedEventOnNotifyFi
-	[InlineAutoData(NotifyFilters.DirectoryName)]
-	public async Task NotifyFilter_MoveDirectoryOutOfTheWatchedDirectory_ShouldTriggerChangedEventOnNotifyFilters(
-		NotifyFilters notifyFilter, string sourceName, string destinationName)
+	[InlineAutoData(NotifyFilters.DirectoryName, true)]
+	[InlineAutoData(NotifyFilters.DirectoryName, false)]
+	public async Task NotifyFilter_MoveDirectoryOutOfTheWatchedDirectory_ShouldTriggerDeletedEventOnNotifyFilters_OnWindows(
+		NotifyFilters notifyFilter, bool includeSubdirectories, string sourceName, string destinationName)
 	{
 		SkipIfLongRunningTestsShouldBeSkipped();
+		Skip.IfNot(Test.RunsOnWindows);
```
Correcting the event that should be triggered, marking as Windows-only, and testing both true and false for `IncludeSubDirectories`.

vbreuss commented on Source/Testably.Abstractions.Testing/FileSystem/FileSystemWatcherMock.cs:
```diff
@@ -538,28 +548,12 @@ private string TransformPathAndName(
 		}
 
 		name = transformedName;
-		if (!_fileSystem.Path.IsPathRooted(Path))
-		{
-			string rootedWatchedPath = _fileSystem.Path.GetFullPath(Path);
-			if (path?.StartsWith(rootedWatchedPath, _fileSystem.Execute.StringComparisonMode) ==
-			    true)
-			{
-				path = _fileSystem.Path.Combine(Path, path.Substring(rootedWatchedPath.Length));
-			}
-		}
-
```
This was added on purpose.
During testing of #866, I stumbled upon the case that you initialize the file system watcher with a relative directory (e.g. "watched") instead of one with a full path and the returned paths were also relative. Currently the tests fail on the real file system:
<img width="1478" height="287" alt="Image" src="https://github.com/user-attachments/assets/46336279-b528-463f-9ba0-9702020393b5" />

vbreuss commented on Tests/Testably.Abstractions.Tests/FileSystem/FileSystemWatcher/NotifyFiltersTests.cs:
```diff
@@ -682,18 +684,61 @@ public async Task NotifyFilter_MoveDirectoryOutOfTheWatchedDirectory_ShouldTrigg
+	[Theory]
+	[InlineAutoData(NotifyFilters.DirectoryName, true)]
+	[InlineAutoData(NotifyFilters.DirectoryName, false)]
+	public async Task NotifyFilter_MoveDirectoryInToTheWatchedDirectory_ShouldTriggerCreatedEventOnNotifyFilters_OnWindows(
```
In [your comment](https://github.com/Testably/Testably.Abstractions/issues/865#issuecomment-3299468675) you also mentioned files. Maybe we should also add test cases when files are moved in/out of the watched directory?

vbreuss commented on Source/Testably.Abstractions.Testing/FileSystem/FileSystemWatcherMock.cs:
```diff
@@ -353,17 +353,27 @@ private bool MatchesFilter(ChangeDescription changeDescription)
-				return changeDescription.ChangeType == WatcherChangeTypes.Renamed &&
-				       changeDescription.OldPath?.StartsWith(fullPath,
-					       _fileSystem.Execute.StringComparisonMode) == true;
+				if (!(changeDescription.ChangeType == WatcherChangeTypes.Renamed &&
+				      changeDescription.OldPath?.StartsWith(fullPath,
```
```suggestion
				if (!changeDescription.ChangeType == WatcherChangeTypes.Renamed ||
				      !changeDescription.OldPath?.StartsWith(fullPath,
```

I try to avoid negated braces in if-conditions, as I had to read this multiple times ("why don't you check that the old path is **not** in the watched directory?") before understanding it...

vbreuss commented on Source/Testably.Abstractions.Testing/FileSystem/FileSystemWatcherMock.cs:
```diff
@@ -353,17 +353,27 @@ private bool MatchesFilter(ChangeDescription changeDescription)
 			_fileSystem.Execute.StringComparisonMode))
 		{
-			return false;
+			if (!(changeDescription.ChangeType == WatcherChangeTypes.Renamed &&
+			      string.Equals(
```
```suggestion
			if (changeDescription.ChangeType != WatcherChangeTypes.Renamed ||
			    !string.Equals(
```

Same here; I think the inteded logic reads more clearly.

vbreuss reviewed (changes requested):
Thanks for the PR, @stogle. I really like the general approach and have some suggestions for improvement.
I also explained the change I made in #866 that you deleted that currently [fail the build](https://github.com/Testably/Testably.Abstractions/actions/runs/17812028987/job/50645952051?pr=869) on the real file system.

stogle commented:
Thanks for the review @vbreuss. I'm not surprised that some changes are needed. I will look at it again tomorrow. I just saw the build failure, but I'm sure that all the tests passed locally before I pushed? Anyway, I'll figure it out.

vbreuss commented:
> Thanks for the review @vbreuss. I'm not surprised that some changes are needed. I will look at it again tomorrow. I just saw the build failure, but I'm sure that all the tests passed locally before I pushed? Anyway, I'll figure it out.

@stogle: Please note the [TestSettings](https://github.com/Testably/Testably.Abstractions/tree/main/Tests/Settings/Testably.Abstractions.TestSettings) project:
It allows some tests to be skipped automatically during local development (per default no tests against the real file system are executed locally). In order for your tests to run locally also against the real file system you have to either switch to Release mode or execute the [Testably.Abstractions.TestSettings.RealFileSystemTests.EnableAlways](https://github.com/Testably/Testably.Abstractions/blob/main/Tests/Settings/Testably.Abstractions.TestSettings/RealFileSystemTests.cs#L45) test once explicitely.
The NotifyFiltersTests even then are skipped due to the `SkipIfLongRunningTestsShouldBeSkipped()`, so you also need to enable them in the settings [the same way](https://github.com/Testably/Testably.Abstractions/blob/main/Tests/Settings/Testably.Abstractions.TestSettings/LongRunningTests.cs#L55)!

stogle commented:
@vbreuss I've reinstated the code that I removed, refactored the logic in `MatchesFilter` and `TriggerRenameNotification` to improve readability, and added tests for moving files in to, and out of the watched path. But I'm having trouble getting all the "move in" and "move out" tests to pass on all file systems, and it seems to be related to the issue you found with relative paths. Could you help me with this?

vbreuss commented:
> @vbreuss I have a theory about these hacks around constructing `FileSystemEventArgs`, although I'm very new to this codebase so I may be wrong. In [System.IO.FileSystemWatcher](https://github.com/dotnet/runtime/blob/9d5a6a9aa463d6d10b0b0ba6d5982cc82f363dc3/src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs), when they construct `FileSystemEventArgs` or `RenameEventArgs`, they pass in the watcher's directory, not the directory of the file that changed. The constructor then combines them to set the `FullPath` property.
> 
> If `FileSystemWatcherMock` did the same thing in `ToFileSystemEventArgs` and `TryMakeRenamedEventArgs` could some of the hacks be avoided?

@stogle: You are right 😄 :
The hacks can be avoided when simulating the native operating system, but not when simulating a different operating system. But thanks to your input I created another refactor PR (#871) that will eliminate the hacks unless you simulate another OS.

stogle commented:
Ah, I get it now. That explains the failures I was seeing when I tried removing them myself. And your new PR not only improves performance but makes it clearer when and why the hacks are needed.
